aisutd/membership-portal
# feat/data-persistance

### Summary

Improve auth flow and setup connection to DynamoDB for reading / writing user session state information.

### Changes
 - Add Cognito user groups for `Members` and `Officers`
 - Add Lambda function that executes as a Cognito post confirmation trigger to selectively add people to the appropriate groups
 - Add CloudWatch log groups to monitor lambda function execution
 - People with an `@aisociety.io` email get auto-added to the `Officers` group
 - People on the AIS tech team get auto-added to the `Members` group. Note: This is hard coded as an array of emails to validate against. In the next incremental update it should read from a dynamodb table with a list of all the members. 
 - Add custom next-auth session callback to insert user `sub` into session token. 
 - Setup AWS DynamoDB to use as session database. Setup to store Next-Auth Sessions & Cognito Sessions. Most important data stored here is the user's `accessToken`, `refreshToken` & `idToken`. 
 - Setup AWS SDK in membership portal repo
 - Setup scripts to fetch `accessToken` from DynamoDB after session authentication
 - Setup `recoiljs` as a global state management service
 - Setup base types and interfaces for API endpoints
 - Created mock api / recoil state to store `accessToken` and log to console for testing

### Testing
 - Update `.env.local` with new environment variables for AWS access. Use the format below for the environment variables. To get an AWS access key & secret key go to AWS IAM & create a new user with programmatic access (you'll see an example user `harshasrikara` already present). Make sure to grant your new user full admin permissions. Once that's complete you'll be able to copy the access & secret keys for that user and add them to the environment variables. 

Note: For prod deployments, we'll have a dedicated user that contains only the relevant permissions required for initial setup since the signed in user's Cognito access token will be used for all subsequent api calls.

```
NEXT_AUTH_AWS_ACCESS_KEY=
NEXT_AUTH_AWS_SECRET_KEY=
NEXT_AUTH_AWS_REGION=us-east-2
````

 - Start the application with `npm run dev`
 - Sign in
 - Check browser console & should see data along these lines printed

```js
{
     access_token:  "eyJr...",
     message: "success",
     provider_sub: "fd9c36c5-4524-420a-8f33-c892886ef367",
     status: true
}
```

 - If signing in with your `@aisociety.io` account check the AWS Cognito Dashboard & verify that the user was successfully added to the `Members` and `Officers` permission groups.

Note: To be added to the appropriate groups delete your existing Cognito account & remove your entry in DynamoDB. Then sign up for a new account. The logic to determine what roles a user gets added to only executes upon new user sign up.
